src/Rings/mpoly-affine-algebras.jl

```diff
@@ -151,6 +152,11 @@ function monomial_basis(A::MPolyQuoRing)
   return gens(MPolyIdeal(base_ring(I), si))
 end
 
+function monomial_basis(R::MPolyRing, I::MPolyIdeal)
+  base_ring(I) == R || error("ideal does not belong to the correct ring")
```

Comment:
Looking at this I was wondering "Why does one have to specify `R` at all? Why not just this:"
```suggestion
function monomial_basis(I::MPolyIdeal)
  R = base_ring(I)
```

Perhaps the answer is "because this is not really a "monomial basis of `I` so we want to avoid the impression that is by giving two arguments". I can see why one would want to avoid writing down a quotient "just" to get this monomial basis. But it still feels at odds with fundamental OSCAR principles (where we like to avoid such tricks). And in any case, you *still* compute `quo` in here, so there is nothing being saved here...

Mathematically I think this is about finding a monomial basis of a complement of `I` in `R`. So an alternatively would be to do something like this:
```suggestion
function monomial_basis_of_complement(I::MPolyIdeal)
  R = base_ring(I)
```

And then one could go a step further and move most of the body of the `monomial_basis(A::MPolyQuoRing)` method in here, as it only uses `I`, not the quotient ring?

Reply:
What we compute here is a monomial basis of a quotient ring A = R/I, say. As does @fingolfin, I have already problems to understand why we should split this up like in `monomial_basis(R, I)` instead of `monomial_basis(A)` (this may already mislead the unexperienced user). Using just `monomial_basis(I)` makes it even worth. We do compute a monomial basis of A, not of (R, I), and certainly not of I.

Sorry, but `monomial_basis_of_complement(I)` does not make any mathematical sense.

In sum, I suggest to forget this idea of splitting up. In particular, I am against splitting up in cases where we already have a working `monomial_basis(A)`.

Reply:
> Perhaps the answer is "because this is not really a "monomial basis of `I` so we want to avoid the impression that is by giving two arguments". I can see why one would want to avoid writing down a quotient "just" to get this monomial basis. But it still feels at odds with fundamental OSCAR principles (where we like to avoid such tricks). And in any case, you _still_ compute `quo` in here, so there is nothing being saved here...

This is indeed the reason. `monomial_basis(I)` is absolutely not the way to go w.r.t. Oscar principles. But in particular in the context of germs and of ideal sheaves, you internally usually have an ideal  in your hands and not a quotient ring when you find yourself in a setting to call monomial_basis.

To the remark with the `quo`: The lines here are not the key contribution of this PR. This is only the compatibility change, which currently does not touch the old code (as I do not know whether we are allowed to change this). The new contribution is in the local case (other file below) and there the `quo` is not computed, but `monomial_basis(R,I)` is the inner worker, whereas `monomial_basis(quo...)` is the outside appearance. 
By the way, I would like to change the non-local case in the same way, if we are allowed to.

> 
> Mathematically I think this is about finding a monomial basis of a complement of `I` in `R`. So an alternatively would be to do something like this: 
> And then one could go a step further and move most of the body of the `monomial_basis(A::MPolyQuoRing)` method in here, as it only uses `I`, not the quotient ring?
NO, we are not looking at the complement. We are looking at the number of elements(=classes mod I) of the factorring R/I. I could live with monomial_basis_of_quo(I), but how do you type-dispatch on the type of the localization of R with this notation?

Reply:
By the way: `quo(R,I)` also contains the R, even though it is not necessary. We simply tried to stay in tune with `quo` (thinking that it would certainly follow OSCAR's principles).

Reply:
(Since this is not returning a basis of $R/I$, but a lifted basis of $R/I$ to $R$, it is in fact a basis for the complement of $I$, complement as a $K$-subspace of $R$.

Reply:
@wdecker: nobody wants to get rid of  `monomial_basis(A)`! This is the way the average user should use it! 

It is just a question of having a more internal, but documented variant for specialists (e.g. scripting, programming), which avoids creating the `quo` specifically to call monomial_basis which in turn immediately passes back to the modulus. This change is not done here yet. It is only done in the local case below. (But I would like to do it here as well.)

Reply:
If you document it, the user will find it. If it is internal, you may also put an _ in front. Also, you compute the quo anyway. So what do you gain?. Typing monomial_basis(R, I) instead of monomial_basis(quo(R, I)[1])? But, if it is important to you, I will not stand in your way.

Reply:
I see this as an ok to put the work into the internal monomial_basis(R,I) and let the outside one monomial_basis(A) call it, like in the local case.